core/src/main/java/org/elasticsearch/action/search/AbstractSearchAsyncAction.java

```diff
@@ -119,30 +121,33 @@ public void start() {
 
     void performFirstPhase(final int shardIndex, final ShardIterator shardIt, final ShardRouting shard) {
         if (shard == null) {
+            // TODO upgrade this to an assert...
             // no more active shards... (we should not really get here, but just for safety)
             onFirstPhaseResult(shardIndex, null, null, shardIt, new NoShardAvailableActionException(shardIt.shardId()));
         } else {
-            final DiscoveryNode node = nodeIdToDiscoveryNode.apply(shard.currentNodeId());
-            if (node == null) {
-                onFirstPhaseResult(shardIndex, shard, null, shardIt, new NoShardAvailableActionException(shardIt.shardId()));
-            } else {
+            try {
+                final Transport.Connection connection = nodeIdToConnection.apply(shard.currentNodeId());
                 AliasFilter filter = this.aliasFilter.get(shard.index().getUUID());
                 assert filter != null;
 
                 float indexBoost = concreteIndexBoosts.getOrDefault(shard.index().getUUID(), DEFAULT_INDEX_BOOST);
                 ShardSearchTransportRequest transportRequest = new ShardSearchTransportRequest(request, shardIt.shardId(), shardsIts.size(),
                     filter, indexBoost, startTime());
-                sendExecuteFirstPhase(node, transportRequest , new ActionListener<FirstResult>() {
-                        @Override
-                        public void onResponse(FirstResult result) {
-                            onFirstPhaseResult(shardIndex, shard.currentNodeId(), result, shardIt);
-                        }
-
-                        @Override
-                        public void onFailure(Exception t) {
-                            onFirstPhaseResult(shardIndex, shard, node.getId(), shardIt, t);
-                        }
-                    });
+                sendExecuteFirstPhase(connection, transportRequest, new ActionListener<FirstResult>() {
+                    @Override
+                    public void onResponse(FirstResult result) {
+                        onFirstPhaseResult(shardIndex, shard.currentNodeId(), result, shardIt);
+                    }
+
+                    @Override
+                    public void onFailure(Exception t) {
+                        onFirstPhaseResult(shardIndex, shard, connection.getNode().getId(), shardIt, t);
+                    }
+                });
+            } catch (ConnectTransportException | IllegalArgumentException ex) {
+                // we are getting the connection early here so we might run into nodes that are not connected. in that case we move on to
+                // the next shard.
```

Comment:
this one is different since we are now using `connection` and not `DiscoveryNode`. These exceptions are thrown earlier now so we have to handle them. It's instead of the null value returned previously